### PR TITLE
feat: Add wrangler pages project list

### PR DESCRIPTION
**Pull request by sidharthachatterjee**
This adds a `wrangler pages project list` command that lets you list your Cloudflare Pages for an account. 

It reuses the `requireAuth` helper (which I've moved into utils) to render an account selector. 

Here is what it renders:

Screenshot 2022-04-04 at 16 40 08

### CODE REVIEW

File: packages/wrangler/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { loginOrRefreshIfRequired, getAccountId } from "./user";
+
+import type { Config } from "./config";
+
+/**
+ * Ensure that a user is logged in, and a valid account_id is available.
+ */
+export async function requireAuth(
```
Comment (threepointone) [inline packages/wrangler/src/utils.ts:8]: Move this into user.tsx instead

**Review by threepointone — COMMENTED** [review]
Could you add some tests as well?

File: packages/wrangler/src/utils.ts (hunk 1 above)
Comment (sidharthachatterjee) [reply packages/wrangler/src/utils.ts:8]: Makes sense. On it.

**Comment by GregBrimble** [comment]
Can we get pagination or search on this project list as well? Quite possible people have more than whatever we serve by the default.

File: packages/wrangler/src/pages.tsx [hunk 2]
```diff
@@ -1059,6 +1067,33 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
...
+          const accountId = await requireAuth(config);
+          let projects = await fetchResult<any>(
+            `/accounts/${accountId}/pages/projects`
+          );
+
+          projects = projects.map((project: any) => {
+            return {
+              "Project Name": project.name,
+              "Project Domains": `${project.domains.join(", ")}`,
+              Git: !!project.source,
```
Comment (GregBrimble) [inline packages/wrangler/src/pages.tsx:1090]:
```suggestion
              "Git Provider": project.source ? "✅ Yes" : "❌ No",
```

File: packages/wrangler/src/pages.tsx [hunk 3]
```diff
@@ -1059,6 +1067,33 @@ export const pages: BuilderCallback<unknown, unknown> = (yargs) => {
...
+              "Last Modified": project.latest_deployment.modified_on,
```
Comment (GregBrimble) [inline packages/wrangler/src/pages.tsx:1091]: Maybe want to use the same timeago util as you have in the deployment list command?

Comment (sidharthachatterjee) [reply packages/wrangler/src/pages.tsx:1091]: Oh yeah, forgot about that here for some reason 😅 Good catch!

File: packages/wrangler/src/pages.tsx [hunk 4]
```diff
@@ -1067,3 +1113,31 @@ const invalidAssetsFetch: typeof fetch = () => {
...
+  const pageSize = 10;
+  let page = 1;
+  const results = [];
+  while (results.length % pageSize === 0) {
+    const json: Array<Project> = await fetchResult(
+      `/accounts/${accountId}/pages/projects`,
+      {},
+      new URLSearchParams({
+        per_page: pageSize.toString(),
+        order: "title",
```
Comment (GregBrimble) [inline packages/wrangler/src/pages.tsx:1131]: The only params this endpoint accepts are `page`, `per_page` and `name`.

Comment (GregBrimble) [reply packages/wrangler/src/pages.tsx:1131]: Order is fixed as last modified desc.

Comment (sidharthachatterjee) [reply packages/wrangler/src/pages.tsx:1131]: Fun. Thanks for pointing that out!

File: packages/wrangler/src/__tests__/pages.test.ts [hunk 5]
```diff
@@ -54,4 +57,76 @@ describe("subcommand implicit help ran on incomplete command execution", () => {
...
+    });
+    function mockListRequest(namespaces: unknown[]) {
+      const requests = { count: 0 };
+      setMockResponse(
+        "/accounts/:accountId/pages/projects",
+        ([_url, accountId], init, query) => {
+          requests.count++;
+          expect(accountId).toEqual("some-account-id");
+          expect(query.get("per_page")).toEqual("10");
+          expect(query.get("order")).toEqual("title");
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages.test.ts:75]: Removed

File: packages/wrangler/src/__tests__/pages.test.ts [hunk 6]
```diff
@@ -54,4 +57,76 @@ describe("subcommand implicit help ran on incomplete command execution", () => {
...
   });
+
+  describe("project list", () => {
+    mockAccountId();
+    mockApiToken();
+
+    afterEach(() => {
+      unsetAllMocks();
+    });
+    function mockListRequest(namespaces: unknown[]) {
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages.test.ts:68]:
```suggestion
    function mockListRequest(projects: unknown[]) {
```

File: packages/wrangler/src/__tests__/pages.test.ts [hunk 7]
```diff
@@ -54,4 +57,76 @@ describe("subcommand implicit help ran on incomplete command execution", () => {
...
+        "/accounts/:accountId/pages/projects",
+        ([_url, accountId], init, query) => {
+          requests.count++;
+          expect(accountId).toEqual("some-account-id");
+          expect(query.get("per_page")).toEqual("10");
+          expect(query.get("page")).toEqual(`${requests.count}`);
+          expect(init).toEqual({});
+          const pageSize = Number(query.get("per_page"));
+          const page = Number(query.get("page"));
+          return namespaces.slice((page - 1) * pageSize, page * pageSize);
```
Comment (GregBrimble) [inline packages/wrangler/src/__tests__/pages.test.ts:80]:
```suggestion
          return projects.slice((page - 1) * pageSize, page * pageSize);
```

**Review by GregBrimble — APPROVED** [review]

**Pull request merged**